comictalker/comiccacher.py

```diff
@@ -367,14 +346,15 @@ def get_volume_issues_info(self, volume_id: int, source_name: str) -> list[CVIss
 
             # now process the results
             for row in rows:
-                record = CVIssuesResults(
+                record = ComicIssue(
                     id=row[1],
                     name=row[2],
                     issue_number=row[3],
                     site_detail_url=row[4],
                     cover_date=row[5],
-                    image=CVImage(super_url=row[6], thumb_url=row[7]),
+                    image=row[6],
                     description=row[8],
+                    volume={"id": volume_id, "name": ""},  # Name isn't recorded
```

Comment:
This needs to be `get_search_results` because `Volumes` table isn't filled yet. Is there a reason why the `Volumes` table is there if it reproduces (with less information) the same data that's in `VolumeSearchCache`? 

I've altered `get_search_results` to accept an id and added a bool to prevent the purge due to the DB being locked.

Reply:
`VolumeSearchCache` has one key piece of data and that is the search term. If you wanted to change it to be just 3 columns `search,volume_id,source_id` it would perform the same function. That's what databases normally do they link everything by ID's and you make separate calls to get each bit of info. So in this case you would insert all the volumes into it's own table and then insert a row for each search result that just links to the full data in the volumes table.

Reply:
I'd guess the reason it's how it currently is is because a JOIN was unwanted? I don't know if you inherited the DB layout? It may be possible it's faster to duplicate the data rather than using a JOIN? On such small data sets that doesn't seem likely?

Anyway, do we want to mess with it now or leave it as is? The DB layout might need some changes when another talker (MU) comes into to play so could leave it as a TODO?

Reply:
SQLite performance is not really applicable we're dealing with such a small amount of data that even if you tried to it would be difficult to get a query to take even 0.5s.
Most likely it was just simpler when it was made to use two completely separate tables.

As the cache is only actually used by a Talker and not by anything else (actually the image fetcher uses it I think but that's only related to images), it is really more of "please cache your requests when possible, we don't want to put any more stress than needed on the servers, here's an option if you don't want to make one yourself"

Let's change it here because unless it's for speed and designed carefully it's better to not have duplicate data, duplicate data can have issues with being inconsistent and that can be fixed by only having one place to store it.

Reply:
Change `image` and `image_thumb` to `image_url` and `image_thumb_url` so it's obvious that it's a url and not the actual image

Update the volumes and search tables to this
```python
            cur.execute(
                "CREATE TABLE VolumeSearchCache("
                + "search_term TEXT,"
                + "id INT NOT NULL,"
                + "timestamp DATE DEFAULT (datetime('now','localtime')),"
                + "source_name TEXT NOT NULL)"
            )

            cur.execute(
                "CREATE TABLE Volumes("
                + "id INT NOT NULL,"
                + "name TEXT,"
                + "publisher TEXT,"
                + "count_of_issues INT,"
                + "start_year INT,"
                + "image_url TEXT,"
                + "aliases TEXT,"  # Newline separated
                + "description TEXT,"
                + "timestamp DATE DEFAULT (datetime('now','localtime')), "
                + "source_name TEXT NOT NULL,"
                + "PRIMARY KEY (id, source_name))"
            )
```

Change add_search_results to:
```python
    def add_search_results(self, source_name: str, search_term: str, ct_search_results: list[ComicVolume]) -> None:
        con = lite.connect(self.db_file)

        with con:
            con.text_factory = str
            cur = con.cursor()

            # remove all previous entries with this search term
            cur.execute(
                "DELETE FROM VolumeSearchCache WHERE search_term = ? AND source_name = ?",
                [search_term.casefold(), source_name],
            )

            # now add in new results
            for record in ct_search_results:

                cur.execute(
                    "INSERT INTO VolumeSearchCache "
                    + "(source_name, search_term, id) "
                    + "VALUES(?, ?, ?)",
                    (
                        source_name,
                        search_term.casefold(),
                        record["id"],
                    ),
                )

                data = {
                    "source_name": source_name,
                    "timestamp": datetime.datetime.now(),
                }
                data.update(record)
                self.upsert(cur, "volumes", data)
```

and simplify get_search_results to:
```python
    def get_search_results(
        self, source_name: str, search_term: str
    ) -> list[ComicVolume]:
        results = []
        con = lite.connect(self.db_file)
        with con:
            con.text_factory = str
            cur = con.cursor()

            cur.execute(
                "SELECT * FROM VolumeSearchCache INNER JOIN Volumes on"
                " VolumeSearchCache.id=Volumes.id AND VolumeSearchCache.source_name=Volumes.source_name"
                " WHERE search_term=? AND VolumeSearchCache.source_name=?",
                [search_term.casefold(), source_name],
            )

            rows = cur.fetchall()
            # now process the results
            for record in rows:
                result = ComicVolume(
                    id=record[4],
                    name=record[5],
                    publisher=record[6],
                    count_of_issues=record[7],
                    start_year=record[8],
                    image_url=record[9],
                    aliases=record[10],
                    description=record[11],
                )

                results.append(result)

        return results
```


get_volume_info then needs the new columns added:
```python
    def get_volume_info(self, volume_id: int, source_name: str) -> ComicVolume | None:
        result: ComicVolume | None = None

        con = lite.connect(self.db_file)
        with con:
            cur = con.cursor()
            con.text_factory = str

            # purge stale volume info
            a_week_ago = datetime.datetime.today() - datetime.timedelta(days=7)
            cur.execute("DELETE FROM Volumes WHERE timestamp  < ?", [str(a_week_ago)])

            # fetch
            cur.execute(
                "SELECT * FROM Volumes"
                " WHERE id=? AND source_name=?",
                [volume_id, source_name],
            )

            row = cur.fetchone()

            if row is None:
                return result

            # since ID is primary key, there is only one row
            result = ComicVolume(
                id=row[0],
                name=row[1],
                publisher=row[2],
                count_of_issues=row[3],
                start_year=row[4],
                aliases=row[5],
                description=row[6],
            )

        return result
```

And get_volume_issues_info can be simplified to:
```python
    def get_volume_issues_info(self, volume_id: int, source_name: str) -> list[ComicIssue]:
        # get_volume_info should only fail if someone is doing something weird
        volume = self.get_volume_info(volume_id, source_name) or ComicVolume(id=volume_id, name="")
        con = lite.connect(self.db_file)
        with con:
            cur = con.cursor()
            con.text_factory = str

            # purge stale issue info - probably issue data won't change
            # much....
            a_week_ago = datetime.datetime.today() - datetime.timedelta(days=7)
            cur.execute("DELETE FROM Issues WHERE timestamp  < ?", [str(a_week_ago)])

            # fetch
            results: list[ComicIssue] = []

            cur.execute(
                (
                    "SELECT source_name,id,name,issue_number,site_detail_url,cover_date,image_url,thumb_url,description,aliases"
                    " FROM Issues WHERE volume_id=? AND source_name=?"
                ),
                [volume_id, source_name],
            )
            rows = cur.fetchall()

            # now process the results
            for row in rows:
                record = ComicIssue(
                    id=row[1],
                    name=row[2],
                    issue_number=row[3],
                    site_detail_url=row[4],
                    cover_date=row[5],
                    image_url=row[6],
                    description=row[8],
                    volume=volume,
                    aliases=row[9],
                )

                results.append(record)

        return results
```

Reply:
That's great, saved me some thinking, thanks. I added the `image_url` to `get_volume_info` that was missing.